nupic/research/frameworks/continuous_learning/cl_utils.py

```diff
@@ -0,0 +1,256 @@
+    :param pre_batch_callback: Callback function to be called before every batch
+                               with the following parameters: model, batch_idx
+    :type pre_batch_callback: function
+    :param progress_bar: Optional :class:`tqdm` progress bar args.
+                         None for no progress bar
+    :type progress_bar: dict or None
+
+    :return: mean loss for epoch
+    :rtype: float
+    """
+    model.train()
+    # Use asynchronous GPU copies when the memory is pinned
+    # See https://pytorch.org/docs/master/notes/cuda.html
+    async_gpu = loader.pin_memory
+    if progress_bar is not None:
+        loader = tqdm(loader, **progress_bar)
+        # update progress bar total based on batches_in_epoch
+        if batches_in_epoch < len(loader):
+            loader.total = batches_in_epoch
+
+    # Check if training with Apex Mixed Precision
+    # FIXME: There should be another way to check if 'amp' is enabled
+    use_amp = hasattr(optimizer, "_amp_stash")
+    try:
+        from apex import amp
+    except ImportError:
+        if use_amp:
+            raise ImportError(
+                "Mixed precision requires NVIDA APEX."
+                "Please install apex from https://www.github.com/nvidia/apex")
+
+    t0 = time.time()
+    for batch_idx, (data, target) in enumerate(loader):
+        if batch_idx >= batches_in_epoch:
+            break
+
+        num_images = len(target)
+        data = data.to(device, non_blocking=async_gpu)
+        target = target.to(device, non_blocking=async_gpu)
+        t1 = time.time()
+
+        if pre_batch_callback is not None:
+            pre_batch_callback(model=model, batch_idx=batch_idx)
+
+        optimizer.zero_grad()
+        output = model(data)
+        loss = criterion(output, target)
+        del data, target, output
+
+        t2 = time.time()
+        if use_amp:
+            with amp.scale_loss(loss, optimizer) as scaled_loss:
+                scaled_loss.backward()
+        else:
+            loss.backward()
+
+        if len(freeze_params) > 0:
+            with torch.no_grad():
+                for param in freeze_params:
+                    param_module = param[0]
+                    param_indices = param[1]
+                    param_module.grad[param_indices, :] = 0.0
+
+        t3 = time.time()
+        optimizer.step()
+        t4 = time.time()
+
+        if post_batch_callback is not None:
+            time_string = ("Data: {:.3f}s, forward: {:.3f}s, backward: {:.3f}s,"
+                           + "weight update: {:.3f}s").format(t1 - t0, t2 - t1, t3 - t2,
+                                                              t4 - t3)
+            post_batch_callback(model=model, loss=loss.detach(), batch_idx=batch_idx,
+                                num_images=num_images, time_string=time_string)
+        del loss
+        t0 = time.time()
+
+    if progress_bar is not None:
+        loader.n = loader.total
+        loader.close()
+
+
+def train_model(
```

Comment:
The only difference between this function and `train_model` in `model_utils.y` is the `target -= 1`.
Why do you need `target -= 1` on every batch?

Reply:
If you need to change the target you should apply that transformation to the dataset. See https://pytorch.org/docs/stable/torchvision/datasets.html

Reply:
It was part of me rearranging the train/test dataset by class because pytorch's loss function needs actual digits (as opposed to one-hot vectors) and i wanted my network to accommodate the "zero class" too. I did this so that I wouldn't have to change the full batch training/testing sets for interoperability. If there's a cleaner way to do this I'm game.

Reply:
A better way to apply transformations to the `target` component of a dataset is via the dataset `target_transform`. See https://pytorch.org/docs/stable/torchvision/datasets.html
What do you mean "zero class" in this context? There are 12 classes in GSC (unknown, silence, zero, one, two, three, four, five, six, seven, eight, nine) and "zero" is one of them.